File: bindings/lua/examples/byvalue.lua
local source = debug.getinfo(1, "S").source:sub(2)
local dir = source:match("^(.*)/[^/]+$") or "."
package.path = dir .. "/../?.lua;" .. package.path

local urb = require("urb_ffi")
local ffi, mem = urb.ffi, urb.memory

local suf = package.cpath:match("%.dylib") and ".dylib" or ".so"
local lib = ffi.open(dir .. "/libbyvalue" .. suf, ffi.flags.NOW + ffi.flags.LOCAL)

local point_type = ffi.type.struct({
    { name = "x", type = ffi.type.i32() },
    { name = "y", type = ffi.type.i32() },
}, { name = "Point" })

local point_schema = {
    { name = "x", type = "i32" },
    { name = "y", type = "i32" },
}

local pptr = mem.alloc(mem.struct_sizeof(point_schema))
local point = mem.view(pptr, point_schema)
point.x = 7
point.y = 11

local sum_sig = ffi.type.func(ffi.type.i32(), { point_type }, { name = "sum_point" })
local sum = ffi.bind(ffi.sym(lib, "sum_point"), sum_sig)
print("sum_point", sum(point))

local swap_sig = ffi.type.func(point_type, { point_type }, { name = "swap_point" })
local swap = ffi.bind(ffi.sym(lib, "swap_point"), swap_sig)
local swapped_ptr = swap(point)
local swapped = mem.view(swapped_ptr, point_schema)
print("swapped", swapped.x, swapped.y)

local cb_type = ffi.type.func(ffi.type.i32(), { point_type }, { name = "cb" })
local js_cb = ffi.callback(cb_type, function(pt)
    print("callback got", pt.x, pt.y)
    return pt.x * 100 + pt.y
end)

local call_cb_sig = ffi.type.func(ffi.type.i32(), { point_type, ffi.type.pointer(cb_type) }, { name = "call_cb" })
local call_cb = ffi.bind(ffi.sym(lib, "call_cb"), call_cb_sig)
print("call_cb returns", call_cb(point, js_cb.ptr))

local mapper_type = ffi.type.func(point_type, { point_type }, { name = "mapper" })
local mapper_cb = ffi.callback(mapper_type, function(pt)
    return {
        x = pt.y + 1,
        y = pt.x + 2,
    }
end)
local map_point_sig = ffi.type.func(point_type, { point_type, ffi.type.pointer(mapper_type) }, { name = "map_point" })
local map_point = ffi.bind(ffi.sym(lib, "map_point"), map_point_sig)
local mapped = mem.view(map_point(point, mapper_cb.ptr), point_schema)
print("mapped", mapped.x, mapped.y)

local op_type = ffi.type.func(ffi.type.i32(), { ffi.type.i32() }, { name = "op" })
local use_op_type = ffi.type.func(ffi.type.i32(), { ffi.type.pointer(op_type), ffi.type.i32() }, { name = "use_op" })
local use_op_cb = ffi.callback(use_op_type, function(op, value)
    local applied = op(value)
    print("op(value)", applied)
    return applied * 2
end)
local call_with_op_sig = ffi.type.func(ffi.type.i32(), { ffi.type.i32(), ffi.type.pointer(use_op_type) }, { name = "call_with_op" })
local call_with_op = ffi.bind(ffi.sym(lib, "call_with_op"), call_with_op_sig)
print("call_with_op returns", call_with_op(10, use_op_cb.ptr))

ffi.close(lib)
mem.free(pptr)

File: bindings/node/examples/byvalue.js
'use strict';

const path = require('path');
const { ffi, memory: mem } = require('..');

// load the helper library we compiled via Makefile
const libName = path.join(__dirname, 'libbyvalue' + (process.platform === 'darwin' ? '.dylib' : '.so'));
const lib = ffi.open(libName, ffi.flags.NOW | ffi.flags.LOCAL);

// define a simple struct layout
const pointType = ffi.type.struct([
    { name: 'x', type: ffi.type.i32() },
    { name: 'y', type: ffi.type.i32() },
]);

// allocate and populate a point
const pptr = mem.alloc(ffi.type.sizeof(pointType));
const point = ffi.global(pptr, pointType);
point.x = 7;
point.y = 11;

// call sum_point(struct Point) -> i32
const sumSig = ffi.type.func(ffi.type.i32(), [pointType], { name: 'sum_point' });
const sum = ffi.bind(ffi.sym(lib, 'sum_point'), sumSig);
console.log('sum_point', sum(point)); // expect 18

// call swap_point(Point) -> Point (by-value return)
const swapSig = ffi.type.func(pointType, [pointType], { name: 'swap_point' });
const swap = ffi.bind(ffi.sym(lib, 'swap_point'), swapSig);
const swappedPtr = swap(point);
const swapped = ffi.global(swappedPtr, pointType);
console.log('swapped', swapped.x, swapped.y);

// callback example: define a JS callback taking Point by value
const cbType = ffi.type.func(ffi.type.i32(), [pointType], { name: 'cb' });
const jsCb = ffi.callback(cbType, (pt) => {
    console.log('callback got', pt.x, pt.y);
    return pt.x * 100 + pt.y;
});
// call C helper call_cb(point, cbptr)
const callCbSig = ffi.type.func(ffi.type.i32(), [pointType, ffi.type.pointer(cbType)], { name: 'call_cb' });
const call_cb = ffi.bind(ffi.sym(lib, 'call_cb'), callCbSig);
console.log('call_cb returns', call_cb(point, jsCb.ptr));

// callback returning Point by value
const mapperType = ffi.type.func(pointType, [pointType], { name: 'mapper' });
const mapperCb = ffi.callback(mapperType, (pt) => ({
    x: pt.y + 1,
    y: pt.x + 2,
}));
const mapPointSig = ffi.type.func(pointType, [pointType, ffi.type.pointer(mapperType)], { name: 'map_point' });
const map_point = ffi.bind(ffi.sym(lib, 'map_point'), mapPointSig);
const mapped = ffi.global(map_point(point, mapperCb.ptr), pointType);
console.log('mapped', mapped.x, mapped.y);

// callback receiving a function pointer argument
const opType = ffi.type.func(ffi.type.i32(), [ffi.type.i32()], { name: 'op' });
const useOpType = ffi.type.func(ffi.type.i32(), [ffi.type.pointer(opType), ffi.type.i32()], { name: 'use_op' });
const useOpCb = ffi.callback(useOpType, (op, value) => {
    const applied = op(value);
    console.log('op(value)', applied);
    return applied * 2;
});
const callWithOpSig = ffi.type.func(ffi.type.i32(), [ffi.type.i32(), ffi.type.pointer(useOpType)], { name: 'call_with_op' });
const call_with_op = ffi.bind(ffi.sym(lib, 'call_with_op'), callWithOpSig);
console.log('call_with_op returns', call_with_op(10, useOpCb.ptr));

ffi.close(lib);
mem.free(pptr);
